pkg/service/oidc/service.go

```diff
@@ -0,0 +1,296 @@
+	const openID4VCIType = "openid4vci-proof+jwt"
+	if headers.Type() != openID4VCIType {
+		return "", errors.Errorf("typ must be set to %q", openID4VCIType)
+	}
+	// - alg: REQUIRED. A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. MUST NOT be none or an identifier for a symmetric algorithm (MAC).
+	allowedAlgs := map[jwa.SignatureAlgorithm]struct{}{
+		jwa.ES256:  {},
+		jwa.ES256K: {},
+		jwa.ES384:  {},
+		jwa.ES512:  {},
+		jwa.EdDSA:  {},
+		jwa.PS256:  {},
+		jwa.PS384:  {},
+		jwa.PS512:  {},
+		jwa.RS256:  {},
+		jwa.RS384:  {},
+		jwa.RS512:  {},
+	}
+	alg, ok := allowedAlgs[headers.Algorithm()]
+	if !ok {
+		return "", errors.Errorf("alg %q is not allowed", alg)
+	}
+
+	// Only one of kid, jwk, or x5c can be present.
+	kid := headers.KeyID()
+	headerJWK := headers.JWK()
+	certChain := headers.X509CertChain()
+
+	if !((kid != "") != (headerJWK != nil) != (certChain != nil)) {
+		return "", errors.New("exactly one of kid, jwk, or x5c must be present")
+	}
+
+	// We'll do verification later, once we've established that the nonce is one we produced.
+	token, err := jwt.ParseString(proof.JWT, jwt.WithVerify(false))
+	if err != nil {
+		return "", errors.Wrap(err, "parsing jwt")
+	}
+
+	nonceRaw, ok := token.Get("nonce")
+	if !ok {
+		return "", ErrNonceNotPresent
+	}
+	nonce, ok := nonceRaw.(string)
+	if !ok {
+		return "", ErrNonceNotString
+	}
+	if !s.isNonceValid(nonce) {
+		return "", ErrNonceDifferent
+	}
+
+	if kid != "" {
+		// - kid: CONDITIONAL. JOSE Header containing the key ID. If the Credential shall be bound to a DID, the kid refers
+		// to a DID URL which identifies a particular key in the DID Document that the Credential shall be bound to.
+		// MUST NOT be present if jwk or x5c is present.
+
+		if err := did.VerifyTokenFromDID(ctx, s.resolver, kid, keyaccess.JWT(proof.JWT)); err != nil {
+			return "", errors.Wrap(err, "verifying")
+		}
+		return kid, nil
+	}
+
+	if headerJWK != nil {
+		// - jwk: CONDITIONAL. JOSE Header containing the key material the new Credential shall be bound to. MUST NOT be
+		// present if kid or x5c is present.
+		return "", util.NotImplementedError
+	}
+
+	if certChain != nil {
+		// - x5c: CONDITIONAL. JOSE Header containing a certificate or certificate chain corresponding to the key used to
+		// sign the JWT. This element MAY be used to convey a key attestation. In such a case, the actual key certificate
+		// will contain attributes related to the key properties. MUST NOT be present if kid or jwk is present.
+		return "", util.NotImplementedError
+	}
+
+	// The Credential Issuer MUST validate that the proof is actually signed by a key identified in the JOSE Header.
+	return "", errors.New("unreachable")
+}
+
+func (s Service) isNonceValid(nonce string) bool {
+	valid, err := totp.ValidateCustom(nonce, s.key.Secret(), time.Now(), s.validateOpts())
+	if err != nil {
+		logrus.WithError(err).Error("Problem validating")
```

Comment:
if there's an err should we return false?
what if valid == true and err == something?